Skip shards when querying constant keyword fields

salvatore-campagna opened the pull request:
When a query like a match phrase query or a wildcard query targets a constant keyword field we can skip the query execution on shards where the query builder is rewritten to a MatchNoneQueryBuilder.

This allows us to save time and resources executing queries on nodes that do not return any result. Constant keyword fields allow rewriting a MatchPhraseQueryBuider to TermsQueryBuilder which, in turn, is rewritten into a MatchNoneQueryBuilder in case a constant keyword field does not match the value specified by the query.

A match phrase query is often executed on constant keyword fields by different Elasticsearch integrations which target multiple indices by means of index patterns or data streams. In a real world scenario it is likely that the index pattern or data stream includes tens or hundreds of backing indices each of them with multiple shards involved. As a result, skipping shards in such scenario might result in better query performance and better cluster resource usage.

Note, anyway, that currently, execution of the pre-filter and the corresponding "can match" phase depends on the overall number of shards involved and on whether there is at least one of them returning a non-empty result.

We do the rewrite operation on the data node in the so called "can match" phase, taking advantage of the fact that, at that moment, we can access the index mapping and extract information about constant keyword fields and their value. Doing this on the coordinator node is not possible due to the unavailability of the index mapping.

Resolves #95541

salvatore-campagna committed:
feature: skip shards when querying constant keyword fields

When a query like a match phrase query or a wildcard query targets
a constant keyword field we can skip the query execution on shards
where the query builder is rewritten to a MatchNoneQueryBuilder.

This allows us to save time and resources executing queries on
nodes that do not return any result. Constant keyword fields allow
rewriting a MatchPhraseQueryBuider to TermsQueryBuilder which,
in turn, is rewritten into a MatchNoneQueryBuilder in case a
constant keyword field does not match the value specified by the
query.

A match phrase query is often executed on constnt keyword fields
by different Elasticsearch integrations which target multiple
indices by means of index patterns or data streams. In a real world
scenario it is likely that the index pattern or data stream includes
tens or hundreds of backing indices. As a result, skipping shards
in such scenario might result in better query prformance and better
cluster resource usage.

Note, anyway, that currently execution of the pre filter and the
corresponding "can match" phase depends on the voerall number of
shards involved and on whether there is at least one of them returning
a non-empty result.

We do the rewrite operation on the data node in the so called
"can match" phase, taking advantage of the fact that, at that
moment, we can access the index mapping and extract information
about constant keyword fields and their value.

salvatore-campagna committed:
Update docs/changelog/96161.yaml

elasticsearchmachine commented:
Hi @salvatore-campagna, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

salvatore-campagna committed:
fix: use just one value for the 'area' field

salvatore-campagna committed:
fix: rename class o match naming conventions

salvatore-campagna committed:
fix: rename class o match naming conventions

salvatore-campagna committed:
fix: prevent possible NPE

salvatore-campagna committed:
fix: prevent possible NPE

salvatore-campagna committed:
test: reduce the number of shards to kick in shard pre-filtering

salvatore-campagna committed:
fix: use variable instead of raw value

salvatore-campagna committed:
fix: do not try to rewrite match none queries

salvatore-campagna committed:
fix: do not skip shards targeted by queries using runtime mappings

jimczi commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1524,6 +1524,22 @@ public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOExcep
+            if (rewritten.query() instanceof MatchNoneQueryBuilder) {
+                return new CanMatchShardResponse(false, null);
+            }
+        }
```
The `MappingAwareRewriteContext` doesn't seem to be needed. Why not creating a new `SearchExecutionContext` when the index service is [available](https://github.com/elastic/elasticsearch/pull/96161/files#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fL1567).
You should also reuse `queryStillMatchesAfterRewrite(request, context)` to ensure that all cases are checked (alias filter and global aggregations).

jimczi replied:
> If we use SearchExecutionContext then also range queries (and maybe others?) could be resolved to MatchNoneQueryBuilder instance.

The initial change already used `SearchExecutionContext` but it sets the searcher to `null`. That's why I said that the special context is not needed. There's some logic to handle `SearchExecutionContext` with a null searcher, we should add some javadocs to explain the use case (rewriting with the mapping only) but it works.

martijnvg replied:
Right, I see this now. Yes, if we pass a null searcher then this should work. 👍

salvatore-campagna replied:
I had to add an additional null check in `isFieldWithinQuery`. That logic uses the reader which might be null in this case and triggers a NPE at some point.

jimczi reviewed (commented):
I looked at the PR because I thought it would be difficult to achieve but looks like we already handle the case where the `SearchExecutionContext` doesn't have a searcher. Nice one!
I left one comment to simplify further.
I also wonder why the optim is disabled when runtime fields are added in the request? 
What are we protecting against?

salvatore-campagna committed:
fix: simplify logic removing MappingAwareRewriteContext

We just need a search execution context that includes the index service
and the mappings to take advantage of query rewriting based on mappings.

salvatore-campagna committed:
fix: prevent possible NPE when searcher is null

When rewriting a query before the searcher is available the search
execution context uses a null searcher which, later on, results in
a NPE.

salvatore-campagna committed:
fix: return intersects instead of disjoint

salvatore-campagna committed:
Merge branch 'main' into feature/95541-avoid-unnecessary-search-idle-shards

salvatore-campagna committed:
fix: we do not axquire the searcher if we pre-filter the shard

The counter counts the number of times a searcher is acquired. Since
we skip the shard we do not always acquire the searcher (if the query
is rewritten to skip the shard). As a result we change the expected
counter value.

salvatore-campagna committed:
test: use a lower number of shards for shard pre-filtering

salvatore-campagna commented on server/src/test/java/org/elasticsearch/search/SearchServiceTests.java:
```diff
@@ -1027,13 +1027,13 @@ public void testCanMatch() throws Exception {
         CountDownLatch latch = new CountDownLatch(1);
         SearchShardTask task = new SearchShardTask(123L, "", "", "", null, Collections.emptyMap());
-        assertEquals(8, numWrapInvocations.get());
+        assertEquals(6, numWrapInvocations.get());
```
I had to change these values because we early exit due to shard filtering and the counter is not incremented.

jimczi commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1532,6 +1532,22 @@ public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOExcep
         assert request.searchType() == SearchType.QUERY_THEN_FETCH : "unexpected search type: " + request.searchType();
+        if (request.source() != null && (request.source().query() instanceof MatchNoneQueryBuilder) == false) {
+            final SearchExecutionContext searchExecutionContext = new SearchExecutionContext(
+                indicesService.indexServiceSafe(request.shardId().getIndex())
```
We should reuse the `indexService` below, using `indexServiceSafe` might not work for PIT if the index was deleted after the creation. Can you move this code [here](https://github.com/elastic/elasticsearch/pull/96161/files#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fL1575)?

salvatore-campagna replied:
The idea of doing it at the beginning was to do a early exit and avoid doing everything that happens after...

jimczi replied:
These operations are cheap (no IO) so it shouldn't be an issue and we want to handle all cases (PIT and normal search)?

martijnvg replied:
I think we need to do this rewrite before we invoked `IndexShard#acquireSearcher(...)` since this marks the shard as search active again (and would allow scheduled refreshes to really be executed). So maybe we should keep the rewrite in the place it was?

salvatore-campagna replied:
If we do it late shards go into 'search active state'. `acquireSearcher` does `markSearcherAccessed `. WHich makes the shard search active which will cause waiting for refresh...which we want to avoid.

jimczi replied:
I see and the PIT case sets the refresh pending to false anyway so no need to change that. Thanks for spotting.

jimczi commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1532,6 +1532,22 @@ public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOExcep
+                    .newSearchExecutionContext(
+                        request.shardId().getId(),
+                        request.shardRequestIndex(),
+                        null,
```
Using `null` here is okish but we should update the javadoc in [SearchExecutionContext](https://github.com/elastic/elasticsearch/blob/29ecff1802ca31fcb210bd54164723eeab94b646/server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java#L685)?

jimczi reviewed (commented):
I left some additional comments. Thanks for the simplification!

salvatore-campagna committed:
fix: move code such that we can reuse the index service

salvatore-campagna committed:
docs: align javadoc explaining usage of null searcher

jimczi commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1580,6 +1564,21 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
+                    if (queryStillMatchesAfterRewrite(request, searchExecutionContext) == false) {
+                        return new CanMatchShardResponse(false, null);
+                    }
+                }
```
As I said in my previous comment, it's safe to do that for all cases (PIT and normal searcher). Please move the logic [here](https://github.com/elastic/elasticsearch/pull/96161/files/0de29feb016ae7516603a6198d427b03ebc35db6..1ce266bd43af65169d1d2b1cbf7def2a8230c12b#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fR1590).
Can you also add a comment explaining why rewriting without a searcher allows to ignore hasRefreshPending?

salvatore-campagna committed:
fix: move skip logic

salvatore-campagna committed:
fix: undo counter changes after moving skip logic

salvatore-campagna committed:
fix: mve the logic in the try catch to avod leaking the searcher

salvatore-campagna committed:
fix: move logic before we mark the shard search active

The whole reason to do this logic is to do it soon enough to avoid
a shard to be marked search active which would also need to wait for
refresh. We can skip the shard before and avoid incurring additional
waiting by skipping right after acquiring the index service (so to
reuse it) but before marking the shard search active.

salvatore-campagna committed:
test: check refresh didn't happen for idle shards

salvatore-campagna committed:
docs: expain the idea behing skipping shards using index mappings

salvatore-campagna committed:
fix: rename method

martijnvg commented on server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleTests.java:
```diff
@@ -0,0 +1,446 @@
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertNoSearchHits;
+import static org.hamcrest.Matchers.equalTo;
+
+public class SearchIdleTests extends ESSingleNodeTestCase {
```
Why was this test renamed?
I think all test cases need to have the suffix `IT` in `internalClusterTest`?

martijnvg commented on server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java:
```diff
@@ -681,7 +681,10 @@ public IndexReader getIndexReader() {
-     *  for instance if this rewrite context is used to index queries (percolation). */
+     * which happens in two cases:
+     * 1. if this rewrite context is used to index queries (percolation)
+     * 2. if we use mapping information to skip shards while doing shards pre-filtering in the 'can match' phase
```
Basically when a `SearchExecutionContext` is used with no indexReader.

javanna replied:
is the second still true given the recent improvements you made? Are there places where we do mappings based can match with a null searcher?

martijnvg reviewed (commented):
In general this looks good. Left a few minor comments.

salvatore-campagna committed:
fix: restore original test name

salvatore-campagna committed:
fix: use multiline if statement

salvatore-campagna committed:
fix: checkstyle error on line too long

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:
```diff
@@ -692,6 +692,10 @@ public Relation isFieldWithinQuery(
         ) throws IOException {
+            if (reader == null) {
+                // NOTE: in case we rewrite a query before we have a reader available
+                return Relation.INTERSECTS;
```
is this uncovering an existing bug that may be triggered by the percolator, or only caused by doing a rewrite before acquiring a searcher?

salvatore-campagna replied:
No bug, just doing a rewrite before we have a searcher available.

javanna commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1596,6 +1608,37 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
+        if (queryStillMatchesAfterRewrite(request, searchExecutionContext) == false) {
+            return new CanMatchShardResponse(false, null);
+        }
+        return null;
```
Perhaps this should return a boolean. If true we will go ahead, otherwise we will return a can match shard response with the flag set to false?

javanna commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1545,13 +1545,21 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
                     readerContext = findReaderContext(request.readerId(), request);
                     releasable = readerContext.markAsUsed(getKeepAlive(request));
                     indexService = readerContext.indexService();
+                    final CanMatchShardResponse canMatchAfterRewrite = canMatchAfterRewriteWithoutSearcher(request, indexService);
```
I am slightly uneasy with calling rewrite yet another time. We do this because acquireSearcher will wake up the shard from idle state and there is no way around it, meaning if we get a searcher we do need to refresh the shard which is precisely what we are trying to avoid?

javanna reviewed (commented):
left a couple of comments

javanna commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1596,6 +1608,37 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
+            indexService.newSearchExecutionContext(
+                request.shardId().getId(),
+                request.shardRequestIndex(),
+                null,
```
I wonder if passing a null searcher warrants a cleanup in a followup. I know it's a thing that may already happen but I am not sure we should build upon that further. The rewrite needs a `QueryRewriteContext`, which though will call `convertToSearchExecutionContext` to call `getFieldType` against it. Maybe the query rewrite context could expose  `getFieldType` that would work when mappings are available (in the data node that has the index), to cover for situations where we want to rewrite purely based on mappings like in this case. I see though that `ConstantFieldType#matches` requires the `SearchExecutionContext` as an argument so the methods needed are more than just `getFieldType` and the change I am suggesting is more involving than it may seem... @romseygeek double checking if you have ideas

romseygeek replied:
I think it's just `getFieldType` and `getIndexSettings()`, both of which can be happily moved into `QueryRewriteContext` I think?  Ideally we could move a couple of more methods in there and remove the abomination that is `convertToSearchExecutionContext()`....

romseygeek replied:
And `indexNamedMatcher`.  And actually we can move `getWriteableRegistry` and `nowInMillis` into `SearchExecutionContext` as they are never used for rewriting, which nearly balances things out.

salvatore-campagna replied:
@romseygeek so if I understand correctly the idea is to move all of that stuff into `QueryRewriteContext` so we don't need a `SearchExecutionContext` at all?

romseygeek replied:
Yes, precisely.  I'm working on a PR now...

javanna replied:
I'd advocate for the needed refactoring to be a prerequisite of the PR. I think it does not make a lot of sense to me to build on the search context with null searcher if we want to rather rely on a cleaner approach that relies on the query rewrite context. Otherwise we add in tech debt that we are going to have to remove straight-away or will stay around for much longer if we don't get to it? What do you folks think?

salvatore-campagna committed:
fix: return a boolean instead of CanMatchShardResponse

salvatore-campagna commented:
@romseygeek @javanna is it ok to do the refactoring (moving things in `QueryRewriteContext` and `SearchExecutionContext` in a followup PR?

salvatore-campagna commented:
@elasticsearchmachine test this please

salvatore-campagna commented:
@martijnvg @javanna @romseygeek I need at leat one approval to proceed with merging.

martijnvg commented on server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleIT.java:
```diff
@@ -223,4 +231,216 @@ public void testSearchIdleStats() throws InterruptedException {
+                )
+                .get()
+        );
+        assertAcked(
```
Can these mappings be added with the create index call?

martijnvg commented on server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleIT.java:
```diff
@@ -223,4 +231,216 @@ public void testSearchIdleStats() throws InterruptedException {
+        final IndicesStatsResponse activeIndexStatsAfter = client().admin().indices().prepareStats(activeIndex).get();
+        Arrays.stream(activeIndexStatsAfter.getShards()).forEach(shardStats -> assertFalse(shardStats.isSearchIdle()));
+
+        assertRefreshStats(idleIndexStatsBefore, idleIndexStatsAfter);
```
I think we should also have an `assertActiveRefreshStats(...)`? This ensures that total of after is higher than before.

salvatore-campagna replied:
I am not sure we can match results shard by shard...

martijnvg commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1596,6 +1605,32 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
+     * to {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} after being rewritten to
+     * {@link org.elasticsearch.index.query.TermQueryBuilder}. Other queries have a similar pattern, especially when using
+     * {@link org.elasticsearch.index.mapper.ConstantFieldType} or unmapped fields, they all rewrite to
+     * {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} (if conditions are met). This allows us to avoid extra work other than
```
Maybe this text of this example can be simplified? Something like: `On such example is a term based query against a constant keyword field. This queries can rewrite themselves to a `MatchNoneQueryBuilder`.

martijnvg commented on x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:
```diff
@@ -0,0 +1,275 @@
+        // NOTE: we need an empty result from at least one shard
+        assertEquals(1, activeShards.size());
+
+        assertRefreshStats(statsResponse, afterStatsResponse);
```
Maybe as in the other test case?

martijnvg commented on x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:
```diff
@@ -0,0 +1,275 @@
+        final List<RefreshStats> refreshStatsBefore = Arrays.stream(before.getShards()).map(x -> x.getStats().refresh).toList();
+        final List<RefreshStats> refreshStatsAfter = Arrays.stream(after.getShards()).map(x -> x.getStats().refresh).toList();
+        assertTrue(
+            refreshStatsBefore.size() == refreshStatsAfter.size()
```
Maybe as in the other test case?

martijnvg commented on x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:
```diff
@@ -0,0 +1,275 @@
+
+        // THEN
+        final IndicesStatsResponse idleIndexStatsAfter = client().admin().indices().prepareStats(idleIndex).get();
+        Arrays.stream(idleIndexStatsAfter.getShards()).forEach(shardStats -> assertTrue(shardStats.isSearchIdle()));
```
Maybe assert that `idleIndexStatsAfter.getShards()` has elements or isn't empty?

martijnvg replied:
Maybe turn this then into a static helper method? So that we do this for other similar assertions too.

martijnvg reviewed (approved):
Left a few minor comment. Otherwise LGTM.

salvatore-campagna committed:
fix: refactor some shard idle tests

salvatore-campagna committed:
fix: rename method argument

salvatore-campagna committed:
javadoc: rephrase comment

salvatore-campagna commented on x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:
```diff
@@ -0,0 +1,326 @@
+            final RefreshStats beforeRefresh = beforeShard.getStats().refresh;
+            assertTrue(afterStatsAsMap.containsKey(beforeShard.getShardRouting()));
+            final RefreshStats afterRefresh = afterStatsAsMap.get(beforeShard.getShardRouting()).getStats().refresh;
+            assertTrue(afterRefresh.getTotal() >= beforeRefresh.getTotal());
```
Here we should need just `>` (not equals) because a refresh should have happened in the meanwhile after executing the search query. Anyway, if I remove the equal the testing refresh stats on the active index fails because the values are the sane before and after. I am trying to figure out why.

javanna commented:
> is it ok to do the refactoring (moving things in QueryRewriteContext and SearchExecutionContext) in a followup PR?

I commented above. It's not ideal, I think. I would go that way only if the suggested refactoring takes long or we don't plan to work on it at the moment, which does not seem to be the case. @salvatore-campagna how would you feel about working towards the direction that Alan pointed you to?

romseygeek commented:
Looking again at some of this, would it be possible to reuse the `CoordinatorRewriteContext` for anything here?  The API could be something like:

```
MyGenericQueryBuilder.java

protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
    CoordinatorRewriteContext crc = queryRewriteContext.convertToCoordinatorRewriteContext();
    if (crc != null) {
        // we're on the coordinator, do something simple
    }
    SearchExecutionContext sec = queryRewriteContext.convertToSearchExecutionContext();
    if (sec != null) {
        // on the shard, ready to execute the query, do the final rewrite
    }
    // on the shard, in the can match phase, do something with mappings and/or metadata but no searcher
}
```

salvatore-campagna committed:
fix: reuse existing createIndex

salvatore-campagna committed:
fix: do not check refresh stats for active shards

salvatore-campagna commented:
> 

Yes I can work on that but I think this is out of the scope of this PR and makes it way too large (also to review). SO I suggest to do it in a followup PR, and I can do it. Ok?

javanna commented:
> Yes I can work on that but I think this is out of the scope of this PR and makes it way too large (also to review).

Thanks! I was not suggesting to make all needed changes as part of this PR. I would suggest opening separate prerequisite PRs that allow us to make the desired change (skipping shards when querying constant keyword fields) without relying on a null searcher in search context. Getting this PR in first looks to me like adding tech debt that we are forced to tackle straightaway, and unless there is urgency to get some fix in asap, I would advocate for taking the clean approach if possible. I am assuming the desired change will still make 8.9.

salvatore-campagna commented:
> > Yes I can work on that but I think this is out of the scope of this PR and makes it way too large (also to review).
> 
> Thanks! I was not suggesting to make all needed changes as part of this PR. I would suggest opening separate prerequisite PRs that allow us to make the desired change (skipping shards when querying constant keyword fields) without relying on a null searcher in search context. Getting this PR in first looks to me like adding tech debt that we are forced to tackle straightaway, and unless there is urgency to get some fix in asap, I would advocate for taking the clean approach if possible. I am assuming the desired change will still make 8.9.

Thanks Luca,

I will create a new issue where I describe what we need to do for the refactoring and keep this PR open. I will work on it.

salvatore-campagna committed:
Merge branch 'main' into feature/95541-avoid-unnecessary-search-idle-shards

salvatore-campagna committed:
refactor: use a QueryRewriteContext instead of SearchExecutionContext

salvatore-campagna committed:
fix: typo in Javadoc

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/part-1  please.

salvatore-campagna committed:
fix: number of acquireSearchSupplier invocations

salvatore-campagna committed:
fix: not using null searcher anymore

salvatore-campagna commented:
@javanna @romseygeek @martijnvg I adjusted this PR after merging #96353

martijnvg commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1595,6 +1604,23 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
+     * making the shard search active and waiting for refreshes. As a result, we only wait for refreshes to happen on shards that have
+     * actual data.
+     */
+    private static boolean canMatchAfterRewrite(final ShardSearchRequest request, final IndexService indexService) throws IOException {
```
Maybe reword: `This allows us to avoid extra work other than making the shard search active and waiting for refreshes.` to `This allows us to avoid extra work for example making the shard search active and waiting for refreshes.`?

martijnvg reviewed (approved):
LGTM

javanna commented on server/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -661,6 +668,46 @@ public SearchExecutionContext newSearchExecutionContext(
+     * Creates a new {@link QueryRewriteContext}.
+     * This class is used to rewrite queries before we are able to get a valid {@link SearchExecutionContext} and
+     * allows us to anticipate rewriting queries before the query is executed on the data node. Not using a full
+     * SearhcExecutionContext allows us to save on query latency and IO operations.
```
SearchExecutionContext is mistyped. Shall we also clarify that this is for cases where access to the index is not required as we can make decisions based on mappings alone? Also, expand on what saves query latency and IO operations? e.g. not pulling a searcher removes the cost of pulling a searcher as well as the associated cost of refreshing idle shards.

javanna commented on x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:
```diff
@@ -0,0 +1,319 @@
+        assertEquals(refreshStatsBefore.size(), refreshStatsAfter.size());
+        assertTrue(refreshStatsAfter.containsAll(refreshStatsBefore));
+        assertTrue(refreshStatsBefore.containsAll(refreshStatsAfter));
+    }
```
It is odd that we don't have unit tests, yet I see that SearchServiceTests that already exists is also an `ESSingleNodeTestCase`. You may want to simplify your test taking inspiration from it, in that it pulls the search service from the node and calls canMatch directly against it.

javanna reviewed (approved):
I left a couple of comments, especially one on testing, LGTM otherwise.

salvatore-campagna committed:
docs: uodate javadocs

salvatore-campagna committed:
test: can match on matching and non matching field value

salvatore-campagna committed:
fix: remove unused variable

salvatore-campagna committed:
Merge branch 'main' into feature/95541-avoid-unnecessary-search-idle-shards

salvatore-campagna committed:
fix: method parameter name

elasticsearchmachine commented:
@salvatore-campagna according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is labelled `release highlight` but the changelog has no `highlight` section
